compiler/rustc_session/src/config/cfg.rs

```diff
@@ -83,6 +87,76 @@ impl<'a, T: Eq + Hash + Copy + 'a> Extend<&'a T> for ExpectedValues<T> {
     }
 }
 
+/// Disallow "builtin" cfgs from the CLI as they produce incoherent behavior
+pub(crate) fn disallow_cfgs(sess: &Session, user_cfgs: &Cfg) {
+    let disallow = |cfg: &(Symbol, Option<Symbol>), controlled_by| {
+        let cfg_name = cfg.0;
+        let cfg = if let Some(value) = cfg.1 {
+            format!(r#"{}="{}""#, cfg_name, value)
+        } else {
+            format!("{}", cfg_name)
+        };
+        sess.psess.opt_span_buffer_lint(
+            UNEXPECTED_BUILTIN_CFGS,
+            None,
```

Comment:
What happens if you use `DUMMY_SP` here instead of `None`?

Reply:
I though putting `DUMMY_SP` as the primary span would be an un-allowed case (resulting in an ICE) but apparently it's only the case when putting dummy spans for suggestions.

Currently putting a dummy span forces our default error emitter to attach the diagnostic to the crate outermost file, which implies having `ERROR` annotations (unsure of their usefulness here).

<details>

```diff
{
    "$message_type": "diagnostic",
    "children": [
        {
            "children": [
            ],
            "code": null,
            "level": "note",
            "message": "config `windows` is only supposed to be controlled by `--target`",
            "rendered": null,
            "spans": [
            ]
        },
        {
            "children": [
            ],
            "code": null,
            "level": "note",
            "message": "see <https://github.com/rust-lang/rust/issues/xxxxx> for more information",
            "rendered": null,
            "spans": [
            ]
        },
        {
            "children": [
            ],
            "code": null,
            "level": "note",
            "message": "`#[deny(unexpected_builtin_cfgs)]` on by default",
            "rendered": null,
            "spans": [
            ]
        }
    ],
    "code": {
        "code": "unexpected_builtin_cfgs",
        "explanation": null
    },
    "level": "error",
    "message": "unexpected `--cfg windows` flag",
    "rendered": "error: unexpected `--cfg windows` flag\n   |\n   = note: config `windows` is only supposed to be controlled by `--target`\n   = note: see <https://github.com/rust-lang/rust/issues/xxxxx> for more information\n   = note: `#[deny(unexpected_builtin_cfgs)]` on by default\n\n",
    "spans": [
+        {
+            "byte_end": 0,
+            "byte_start": 0,
+            "column_end": 1,
+            "column_start": 1,
+            "expansion": null,
+            "file_name": "[..]/rust/tests/ui/cfg/disallowed-cli-cfgs.rs",
+            "is_primary": true,
+            "label": null,
+            "line_end": 1,
+            "line_start": 1,
+            "suggested_replacement": null,
+            "suggestion_applicability": null,
+            "text": [
+            ]
+        }
    ]
}
```

</details>

Conceptually idk if it makes sense to have the diag attached to the crate since at the  point we lint the crate doesn't even exist yet. I also find it slightly better to not use a dummy span in diag since as can be seen here it result in some unexpected behaviours, which is probably fine when it's a mistake or when something went wrong but doesn't seem clean IMO.

Another thing, it's currently not handled by the `annotate_snippet` emitter, which currently completely silently ignores any diag if any span is a dummy span, cf. https://github.com/rust-lang/rust/blob/d371d17496f2ce3a56da76aa083f4ef157572c20/compiler/rustc_errors/src/annotate_snippet_emitter_writer.rs#L139-L142